// @nuxthub/core/dist/runtime/utils/auth.js
import { getHeader, createError } from "h3";
import { $fetch } from "ofetch";
const localCache = /* @__PURE__ */ new Map();
export async function requireNuxtHubAuthorization(event) {
  if (import.meta.dev) {
    return;
  }
  const secretKeyOrUserToken = (getHeader(event, "authorization") || "").split(" ")[1];
  if (!secretKeyOrUserToken) {
    throw createError({
      statusCode: 403,
      message: "Missing Authorization header"
    });
  }
  const projectKey = process.env.NUXT_HUB_PROJECT_KEY;
  const projectSecretKey = process.env.NUXT_HUB_PROJECT_SECRET_KEY;
  if (projectSecretKey && secretKeyOrUserToken === projectSecretKey) {
    return;
  } else if (projectSecretKey && !projectKey) {
    throw createError({
      statusCode: 401,
      message: "Invalid secret key"
    });
  }
  if (projectKey) {
    if (localCache.has(secretKeyOrUserToken)) {
      return;
    }
    await $fetch(`/api/projects/${projectKey}`, {
      baseURL: process.env.NUXT_HUB_URL || "https://admin.hub.nuxt.com",
      method: "HEAD",
      headers: {
        authorization: `Bearer ${secretKeyOrUserToken}`
      }
    });
    localCache.set(secretKeyOrUserToken, true);
    return;
  }
  throw createError({
    statusCode: 401,
    message: "Missing NUXT_HUB_PROJECT_SECRET_KEY environment variable or NUXT_HUB_PROJECT_KEY environment variable"
  });
}
